Implement FWaaS for ASR1k

swagner-de opened the pull request:
- Add FWaaS service provider stub
- Checkin FWaaS config draft
- Implement FWaaS CRUD hooks
- Include FWaaS data in sync_data
- Push updated config snippets
- add subnet, host and port attributes to ACL model
- Implement fwg policies as ACLs on interfaces
- Add policy update to router sync
- Extend API to return gateway port id on router object
- Add boilerplate config for allow all policy
- Add class-map and service-policies for external policies
- Adding zones and zone pair for external policies
- Create interface RII and zone-membership if external policy present
- Register VRF with parameter map for policing
- Implement fake stateful rules
- Generic signature for `is_orphan`
- Implement deletion of FWaaS router specific objects
- Add diff to FWaaS objects
- Pass `should_be_none` all the way through
- Add API to retrieve router centric view of firewall rules
- Include FWAAS objects that are solely owned by the router in cleaner
- Be more certain that connections are marked alive before loopingcalls
- Implement cleanup loop for FWAAS objects not tied to a router
- Move FWAAS object prefixes into const module
- Add FWAAS objects parsing tests

sebageek committed:
Add FWaaS service provider stub

If specified as a service_provider for FIREWALL_V2 the ASR1KFWaaSDriver
will receive all pre/postcommit hooks necessary to implement the
OpenStack FWaasV2 service.

swagner-de committed:
Checkin FWaaS config draft

swagner-de committed:
Implement FWaaS CRUD hooks

swagner-de committed:
Include FWaaS data in sync_data

swagner-de committed:
Push updated config snippets

swagner-de committed:
add subnet, host and port attributes to ACL model

swagner-de committed:
Implement fwg policies as ACLs on interfaces

swagner-de committed:
Add policy update to router sync

fwg policies are shared objects, they can be used by multiple routers.
We map them 1:1 to ACL objects on IOS XE and thus need a method to keep
them updated. We could do that by calling a router sync on any single
router that has a reference to that policy or we just call it on an
arbitrary router on every agent. This is what we implement in this
commit.

swagner-de committed:
Extend API to return gateway port id on router object

In our FWaaS implementation we will most likely bind policies on
external interfaces. In order to tell the customer the external
interface we expand the API to convey this information.

swagner-de committed:
Add boilerplate config for allow all policy

swagner-de committed:
Add class-map and service-policies for external policies

swagner-de committed:
Adding zones and zone pair for external policies

swagner-de committed:
Create interface RII and zone-membership if external policy present

In order for the ZBF feature to replicate the state table, interfaces
need to be assigned to a redundandancy group and receive a Redundant
Interface Identifier (RII) in order to find the same interface on the
remote side. In case we find an external policy on the `Router`, we
mark all interfaces on the router's interface list with
`has_stateful_firewall`. If that is the case, we instantiate the yang
model with RII, redundancy group membership and in case of the interface
being the gateway interface with an zone membership. As this marking is
done post `Interface` instantiation, we make the `_rest_definition`
attribute a dynamic property, so later changes in the `Interface` object
are represented in its `_rest_definition`.

swagner-de committed:
Register VRF with parameter map for policing

swagner-de committed:
Implement fake stateful rules

As we cannot support hardware stateful firewalling, we mimic stateful
behaviour by inserting a set of rules of the beginning of each ACL that
help us letting  through packets that are usually an answer to a sent
out packet.

swagner-de committed:
Implement deletion of FWaaS router specific objects

Zones and ZonePairs are created per VRF that has an external policy
attached. We can remove them straigh away once the external policy was
removed on an VRF/external router interface.

swagner-de committed:
Add diff to FWaaS objects

swagner-de committed:
Pass `should_be_none` all the way through

swagner-de committed:
Add API to retrieve router centric view of firewall rules

FWaaS rules are mapped via policies and firewall groups to ports. Yet
when we need to troubelshoot, we must compare it to a virtual router's
configuration, so we'd need easy access to how the rules would look from
a router interface perspective. This commit adds an API endpoint that
delivers that view, removing some cluttering attributes from the result.

swagner-de committed:
Include FWAAS objects that are solely owned by the router in cleaner

swagner-de committed:
Be more certain that connections are marked alive before loopingcalls

I noticed that the first iterations of all loopingcalls would always
fail. Turns out, `_run_yang_cmd` checks the `alive` condition of the
context, which is not set until `device_check_loop` has run the first
time. To make matters worse, `device_check_loop` is also the last
loopingcall to be triggered, so it's certain that the first iteration
off all previous calls will fail.

I moved the loopingcall to be the first one triggered and also halt
until we see the agents marked alive. I believe that makes sense as no
yang action can be run before this happens either way. As I don't want
to inhibit agent startup alltogether, I still back out after
`nc_timeout`.

swagner-de committed:
Implement cleanup loop for FWAAS objects not tied to a router

Firewall policies can be bound to multiple ports. They are realized by
ACLs, a ClassMap and a ServicePolicy (if the policy is bound to an
external interface). That means, just because a router or a port is
removed from a policy the corresponding objects may still be used from
other routers. Hence it is not removed immediatly but by a cleanup loop,
that checks if anybody on an agent is using that policy and then it will
be removed.

swagner-de committed:
Move FWAAS object prefixes into const module

This prevents cyclic imports as the prefix is needed in the orphan
check, but also needed when creating the object through one of the
higher level l3 classes.

swagner-de committed:
Add FWAAS objects parsing tests

sebageek commented on asr1k_neutron_l3/models/netconf_yang/l3_interface.py:
```diff
@@ -216,6 +251,28 @@ def to_dict(self, context):
+        else:
+            ip[L3Constants.ACCESS_GROUP] = {
+                L3Constants.OUT: {
+                    xml_utils.OPERATION: NC_OPERATION.DELETE
```
Would use a `REMOVE` instead of `DELETE`, or are you sure that the element is present on the device in all cases?

swagner-de replied:
I am not sure that this element is always present, as a matter of fact I'd expect it to be absent when no access-group is defined. However there was now exception for that case yet, which is weird. I'll use `REMOVE` never the less.

sebageek commented on asr1k_neutron_l3/models/netconf_yang/l3_interface.py:
```diff
@@ -216,6 +251,28 @@ def to_dict(self, context):
+        else:
+            ip[L3Constants.ACCESS_GROUP] = {
+                L3Constants.IN: {
+                    xml_utils.OPERATION: NC_OPERATION.DELETE
```
Would use a `REMOVE` instead of `DELETE`, or are you sure that the element is present on the device in all cases?

swagner-de replied:
See previous comment.

sebageek commented on asr1k_neutron_l3/plugins/l3/service_plugins/l3_extension_adapter.py:
```diff
@@ -314,9 +317,42 @@ def get_sync_data(self, context, router_ids=None, active=None, host=None):
+                    policies[fwp_id][f"{direction}_ports"].append(port_id)
+        return policies
+
+    def get_deleted_router_atts(self, context, router_ids):
```
Why did this method get an extra parameter `router_ids` it did not have before? I did not yet look through all commits, just noticed that this was added. It doesn't seem to be used in the method.

swagner-de replied:
I don't know. I'll remove it.

sebageek commented on asr1k_neutron_l3/plugins/l3/service_plugins/l3_extension_adapter.py:
```diff
@@ -383,6 +419,13 @@ def update_router(self, context, id, router):
 
+    @resource_extend.extends([l3_def.ROUTERS])
+    def _extend_router_dict(result_dict, db):
+        if not result_dict.get('external_gateway_info'):
+            return result_dict
+        result_dict['external_gateway_info']['external_port_id'] = db.gw_port.id
+        return result_dict
```
How about
```python
if result_dict.get('external_gateway_info'):
    result_dict['external_gateway_info']['external_port_id'] = db.gw_port.id
return result_dict
```

swagner-de replied:
That makes more sense.

sebageek commented on asr1k_neutron_l3/extensions/asr1koperations.py:
```diff
@@ -231,6 +234,43 @@ def update(self, request, id, body, **kwargs):
+        try:
+            ports = {x['id']: x for x in self.plugin.db.get_router_ports(request.context, id)}
+            if not ports:
+                raise exceptions.HTTPNotFound(detail="Router does not exist or has no ports")
```
This exception is caught by your exception handler in L269.

swagner-de replied:
Is there a nicer way other than adding another handler that just re-raises the exception?
```
except exceptions.HTTPNotFound as e:
    raise e
```

sebageek replied:
As you just reraise it I'd just not catch it and let it handle someone else. If you want to reraise something, you can just do a `raise` without arguments.

sebageek commented on asr1k_neutron_l3/models/neutron/l3/firewall.py:
```diff
@@ -0,0 +1,262 @@
+    def __init__(self, policy_id: str, rules: List[dict]):
+        self.policy_id = policy_id
+        super().__init__(self.get_id_by_policy_id(policy_id))
+        self.rules = copy(self.MIMIC_STATEFUL_RULES)
```
Is a copy enough or do we need a deepcopy?

swagner-de replied:
I believe that would only be needed if the stateful rules were altered. Yet that doesn't happen, there are only ever rules appended to the list.

sebageek replied:
Makes sense, we only need the rule references, but we need a new list so we don't get any extra MIMIC_STATEFUL_RULES. Thanks for explaining.

sebageek commented on asr1k_neutron_l3/models/neutron/l3/firewall.py:
```diff
@@ -0,0 +1,262 @@
+        super().__init__(self.get_id_by_policy_id(policy_id))
+        self.rules = copy(self.MIMIC_STATEFUL_RULES)
+        for rule in rules:
+            if not rule['enabled']:
+                continue  # TODO
+            if rule['ip_version'] != 4:
+                continue  # TODO
```
TODO for this PR or for later? If for later, then an explanation would be nice what's to do here.

swagner-de replied:
This is a TODO for when we add IPv6 support. Since that is still a bit out, I removed my comment.

sebageek commented on asr1k_neutron_l3/models/neutron/l3/firewall.py:
```diff
@@ -0,0 +1,262 @@
+                continue  # TODO
+
+            rule_args = {
+                'action': self.ACTIONS[rule['action']],
```
ever possible to get an action that we haven't mapped?

swagner-de replied:
Not until this part in the database is altered:
https://github.com/sapcc/neutron-fwaas/blob/allow-external-router-interfaces/neutron_fwaas/db/firewall/v2/firewall_db_v2.py#L80

sebageek commented on asr1k_neutron_l3/models/neutron/l3/router.py:
```diff
@@ -85,6 +88,34 @@ def __init__(self, router_info):
                            rd=self.router_atts.get('rd'), routable_interface=self.routable_interface,
                            rt_import=self.rt_import, rt_export=self.rt_export, global_vrf_id=global_vrf_id)
 
+        self.fwaas_conf = list()
+        self.fwaas_external_policies = {'ingress': None, 'egress': None}
+        for name, policy in router_info.get('fwaas_policies', {}).items():
+            if self.gateway_interface.id in policy['ingress_ports'] \
+                    or self.gateway_interface.id in policy['egress_ports']:
+                # This policy will be bound on a external interface, so we need to create
+                # class-map and service-policy
+                if self.gateway_interface.id in policy['ingress_ports']:
+                    self.fwaas_external_policies['ingress'] = name
+                if self.gateway_interface.id in policy['egress_ports']:
+                    self.fwaas_external_policies['egress'] = name
+                self.fwaas_conf.append(firewall.ClassMap(name))
+                self.fwaas_conf.append(firewall.ServicePolicy(name))
+            self.fwaas_conf.append(firewall.AccessList(name, policy['rules']))
+
+        if self.fwaas_external_policies['ingress'] or self.fwaas_external_policies['egress']:
+            # As there are external interfaces policies, we create zones and zone-pairs
+            self.fwaas_conf.append(firewall.Zone(self.router_id))
+            self.fwaas_conf.append(
+                firewall.ZonePairExtEgress(self.router_id, self.fwaas_external_policies['egress']))
+            self.fwaas_conf.append(
+                firewall.ZonePairExtIngress(self.router_id, self.fwaas_external_policies['ingress']))
+            # We also want to link the VRF to a policer so we limit VRFs (by boilerplate)
+            self.fwaas_conf.append(firewall.FirewallVrfPolicer(self.router_id))
+            # Mark all interfaces for stateful firewalling
+            for interface in self.interfaces.all_interfaces:
+                interface.has_stateful_firewall = True
```
[optional] Could be `def _build_fwaas_conf()`

swagner-de replied:
Done.

sebageek commented on asr1k_neutron_l3/neutron/services/service_drivers/asr1k/fwaas_driver.py:
```diff
@@ -0,0 +1,154 @@
+                    notification_directory[(policy_id, x.host)]: x.device_id
+                    for x in self.asr1k_db.get_routers_with_policy(context, policy_id=policy_id)}
+            for (policy_id, agent), router_id in notification_directory.items():
+                LOG.debug("{} {} has changed, policy {} associated. Notifying agent to call update on"
+                          " router_id {} on agent {}".format(entity, id, policy_id, router_id, agent))
```
Would use format strings here (more efficient logging, `"%s %s has changed, ..."`)

swagner-de replied:
Done

sebageek reviewed (commented):
First high-level review pass from my side. Overall I like it!

swagner-de committed:
Move FWAAS related object creation into a dedicated method

sebageek commented on asr1k_neutron_l3/models/netconf_yang/access_list.py:
```diff
@@ -108,6 +118,14 @@ def neutron_router_id(self):
+    def policy_id(self):
+        prefix = 'ACL-FWAAS-'
+        if self.name is not None and self.name.startswith('ACL-FWAAS-'):
+            uuid = self.name.lstrip(prefix)
```
This would strip every char in prefix, disregarding the order you pass it in

sebageek commented on asr1k_neutron_l3/plugins/l3/agents/asr1k_l3_agent.py:
```diff
@@ -358,6 +369,17 @@ def after_start(self):
+                if not context.wait_alive():
+                    LOG.error(f"Connection to {context.host} not alive, still continuing startup")
+                else:
+                    LOG.debug(f"Connection to {context.host} alive")
```
[optional] For logging I'd use `%s`-style formatting, as formatting is only done when the log line would be logged (right loglevel is set)

sebageek commented on asr1k_neutron_l3/models/netconf_yang/class_map.py:
```diff
@@ -0,0 +1,106 @@
+
+    def to_dict(self, context):
+        _C = ClassMapConstants
+        content = OrderedDict()
```
Do we need/want to keep OrderedDicts? Felt to me like the normal dicts do just fine in general and I think since python 3.6 we have insertion-order anyway, at least in cpython.

swagner-de committed:
Use sapcc downstream neutron-fwaas

sebageek commented on asr1k_neutron_l3/neutron/services/service_drivers/asr1k/fwaas_driver.py:
```diff
@@ -0,0 +1,153 @@
+LOG = logging.getLogger(__name__)
+
+
+class ASR1KFWaaSDriver(driver_api.FirewallDriverDB, driver_api.FirewallDriverRPCMixin):
```
Why do we inherint from `FirewallDriverRPCMixin`? From the source it looks like all this does is a) provide a `start_rpc_listeners()` method and b) tries to invoke RPC workers in neutron-api (which looks like an upstream bug, which we also could correct). But as this is the driver we're running inside neutron ourselves and already have set up all the rpc we need I think we can drop the inheritance here.

sebageek commented on asr1k_neutron_l3/common/prometheus_monitor.py:
```diff
@@ -110,6 +110,8 @@ def __setup__(self, host=None, namespace=None, type=L3):
             self._gateways = Gauge('gateways', 'Number of managed gateways', STATS_LABELS, namespace=self.namespace)
             self._floating_ips = Gauge('floating_ips', 'Number of managed floating_ips',
                                        STATS_LABELS, namespace=self.namespace)
+            self.fwaas_clean_duration = Histogram("fwaas_cleaner_duration", "FWaaS cleaner runtime in seconds",
```
[very optional] variable named `clean` variable name vs `cleaner` prometheus metric name

sebageek commented on asr1k_neutron_l3/extensions/asr1koperations.py:
```diff
@@ -231,6 +234,45 @@ def update(self, request, id, body, **kwargs):
 
+class FWAASController(wsgi.Controller):
+    def __init__(self, plugin):
+        super(FWAASController, self).__init__()
```
[optional] since python3 we we can just do a `super().__init__()`. I've started to replace this everywhere I see it when I refactor code.

sebageek commented on asr1k_neutron_l3/plugins/l3/service_plugins/l3_extension_adapter.py:
```diff
@@ -179,6 +181,7 @@ def _check_internal_net_az_hints(self, resource, event, trigger, payload, **kwar
                 raise exc
 
 
+@resource_extend.has_resource_extenders
```
I think I would expect the resource extends in`Asr1KPluginBase` - I think this is where I came out at when I reviewed how this works for the flavor-based nat solution (that we didn't choose). What do you think?

sebageek replied:
...what I meant to say is it could live as part of `ASR1KRouterPlugin`, which is similar to Neutrons `L3RouterPlugin`. But your argument to keep it here is solid, let's keep the more neutron specific stuff in the RouterPlugin and our own implementation in the PluginBase.

sebageek commented on asr1k_neutron_l3/plugins/l3/service_plugins/l3_extension_adapter.py:
```diff
@@ -317,8 +320,41 @@ def get_sync_data(self, context, router_ids=None, active=None, host=None):
+        policies = {}
+
+        for port_id in port_ids:
+            with context.session.begin(subtransactions=True):
```
Use `db_api.CONTEXT_READER.using(context)` here.

Does one session per port make sense or should we move the `for` into the transaction context?

swagner-de replied:
It doesn't really hurt to use an individual transaction, yet since this is called out of the context of a router, I think it makes sense to let the tx span multiple ports.

sebageek commented on asr1k_neutron_l3/plugins/l3/service_plugins/l3_extension_adapter.py:
```diff
@@ -317,8 +320,41 @@ def get_sync_data(self, context, router_ids=None, active=None, host=None):
+                fwg_id = self.db.get_fwg_attached_to_port(context, port_id)
+                if not fwg_id:
+                    continue
+                fwg = self.db.get_firewall_group(context, fwg_id,
+                                                 fields=["egress_firewall_policy_id",
+                                                        "ingress_firewall_policy_id",
+                                                        "name", "id"])
```
[optional] do you think we often fetch the same firewall groups here? Could it make sense to cache them? Might make router sync faster. Might also be a followup.

pep8: indent the two list lines by one space

swagner-de replied:
Potentially. This is called on the ports of a router. I suppose it's not an absurd use case for the internal ports to share the same group. If caching means putting the result in a dict and toss it away once we exit this function, this makes sense.

sebageek commented on asr1k_neutron_l3/plugins/l3/service_plugins/l3_extension_adapter.py:
```diff
@@ -317,8 +320,41 @@ def get_sync_data(self, context, router_ids=None, active=None, host=None):
+                fwg_id = self.db.get_fwg_attached_to_port(context, port_id)
+                if not fwg_id:
+                    continue
+                fwg = self.db.get_firewall_group(context, fwg_id,
+                                                 fields=["egress_firewall_policy_id",
+                                                        "ingress_firewall_policy_id",
+                                                        "name", "id"])
+                for fwp_id, direction in ((fwg["egress_firewall_policy_id"], "egress"),
+                                          (fwg["ingress_firewall_policy_id"], "ingress")):
+                    if fwp_id is None:
+                        continue
+                    if fwp_id not in policies:
+                        name = self.db.get_firewall_policy(context, fwp_id, fields=["name"])["name"]
+                        policies[fwp_id] = {
+                            "name": name,
+                            "ingress_ports": [],
+                            "egress_ports": [],
+                            "rules": self.db._get_policy_ordered_rules(context, fwp_id)
+                        }
+                    policies[fwp_id][f"{direction}_ports"].append(port_id)
```
Why can't we use `self.db.make_firewall_group_dict_with_rules()` here? I only looked 10 seconds at it but from that short look it looked like it does something similar.

swagner-de replied:
It does something similar. It returns the rules that are transitively connected (via a policy) to a firewall group. Yet in our FWaaS implementation we are pushing firewall policies as ACLs on the box. So for us it's important to get the UUID of the policy and the ports and direction the policy is attached. 

We could use `make_firewall_group_dict_with_rules`, but that would entail creating a ACL per firewall group, which would also work, it would just lead to us creating more ACLs with potentially the same content. ACLs and ACEs are limited resources (even though IIRC on the Cat 8500 the limit is so high that this wouldn't create a problem), so I chose the approach requiring less ACLs, ACEs and thus less config changes.

sebageek commented on asr1k_neutron_l3/extensions/asr1koperations.py:
```diff
@@ -231,6 +234,45 @@ def update(self, request, id, body, **kwargs):
+            fw_policies = self.plugin.get_fwaas_policies(request.context, ports.keys())
+            rules = {p: {'port': {}, 'ingress': {}, 'egress': {}} for p in ports.keys()}
+            for policy_id, fw_policy in fw_policies.items():
+                for direction in ['ingress', 'egress']:
```
[optional] I tend to use tuples for things like this. I _think_ they're faster because they're immutable, though it doesn't really matter in this case.

sebageek commented on asr1k_neutron_l3/extensions/asr1koperations.py:
```diff
@@ -231,6 +234,45 @@ def update(self, request, id, body, **kwargs):
+            rules = {p: {'port': {}, 'ingress': {}, 'egress': {}} for p in ports.keys()}
+            for policy_id, fw_policy in fw_policies.items():
+                for direction in ['ingress', 'egress']:
+                    for port in fw_policy[f'{direction}_ports']:
```
[optional] `port` -> `port_id`

sebageek commented on asr1k_neutron_l3/extensions/asr1koperations.py:
```diff
@@ -231,6 +234,45 @@ def update(self, request, id, body, **kwargs):
+                for direction in ['ingress', 'egress']:
+                    for port in fw_policy[f'{direction}_ports']:
+                        rules[port][direction]['policy_id'] = policy_id
+                        rules[port][direction]['rules'] = []
```
[optional] Might be easier to read when we create a dict for this and then at the end assign this to `rules[port][direction]`.

sebageek commented on asr1k_neutron_l3/models/asr1k_pair.py:
```diff
@@ -117,6 +119,14 @@ def _get_version_attr(self, attr_name):
             raise VersionInfoNotAvailable(host=self.context.host, entity=attr_name)
         return getattr(self, attr_name)
 
+    def wait_alive(self):
+        timed_out = time.time() + self.nc_timeout
+        while time.time() < timed_out:
+            if self.alive:
+                return True
+            time.sleep(1)
+        return False
```
[optional] Would this mean that the driver would block without a message on start if the device is not reachable? Depending on the timeout would it maybe make sense to log something every.... 10 seconds or some thing? Or have a `log.debug()` once per iteration?

swagner-de replied:
It logs out that it's waiting for the connections to come alive in the caller. I also added 10s logging now.

sebageek commented on asr1k_neutron_l3/models/netconf_yang/access_list.py:
```diff
@@ -108,6 +118,14 @@ def neutron_router_id(self):
+    @property
+    def policy_id(self):
+        prefix = 'ACL-FWAAS-'
+        if self.name is not None and self.name.startswith('ACL-FWAAS-'):
```
Don't you have a constant for that? ;)

Doesn't really need a `is not None` check imo, `if self.name and ...:` would suffice.

sebageek commented on asr1k_neutron_l3/models/netconf_yang/access_list.py:
```diff
@@ -108,6 +118,14 @@ def neutron_router_id(self):
+            uuid = self.name[len(prefix):]
+            if utils.is_valid_uuid(uuid):
+                return uuid
+
```
[optional] I once read somewhere that if you return something in a method you should make all returns explicit, i.e. in this case add a `return None`. Guess it's more of a style question.

sebageek commented on asr1k_neutron_l3/models/netconf_yang/access_list.py:
```diff
@@ -140,10 +158,22 @@ def update(self, context):
+        # Back out if we were called from the clean_fwaas loop
+        if "all_router_ids" not in kwargs:
+            return False
+
```
```python
all_fwaas_policies = kwargs.get("all_fwaas_policies")
if all_fwaas_policies:  # could even use a walrus operator here ;)
    # fwaas syncloop
    if self.policy_id:
        return self.policy_id not in all_fwaasL-policies
    else:
        return False
    
# ... now let the regular code handle it. alternatively put it in an else.
```

sebageek commented on asr1k_neutron_l3/models/netconf_yang/access_list.py:
```diff
@@ -140,10 +158,22 @@ def update(self, context):
         return context.version_min_17_3 and \
             (self.drop_on_17_3 or self.name.startswith("PBR-") and self.neutron_router_id is not None) or \
-            super(AccessList, self).is_orphan(all_router_ids, all_segmentation_ids, all_bd_ids, context)
+            super(AccessList, self).is_orphan(*args, **kwargs)
```
Context argument is missing, add a `context=context` before `**kwargs`. This is only for completeness, though. In my IPv6 branch (which I'll have to rebase onto all these changes) I have removed some of the 17_4 backwards compatibility stuff. We could also cherry-pick it but I think it might not be worth it. Have a look though if you like at #112

sebageek commented on asr1k_neutron_l3/models/netconf_yang/bulk_operations.py:
```diff
@@ -77,7 +77,7 @@ def get_all_stubs_from_device(cls, context):
             return []
 
-    def is_orphan(self, all_router_ids, all_segmentation_ids, all_bd_ids, context):
+    def is_orphan(self, all_router_ids, *args, **kwargs):
```
We know what arguments this method accepts and we should list them here explicitly. Using `*args, **kwargs*` would allow any argument and we have no check whatsoever.

swagner-de replied:
The way like `is_orphan` is used on other classes, is that it lists the arguments it is actually using explicitly while sending all the other possible arguments to `*args` and `**kwargs` just to get rid of them. This way we can add the number of arguments that we pass to its invocation in the cleaner without having to touch each classes implementation of the same method. So I felt like having less control about the arguments was kind of how this method was used already.

sebageek commented on asr1k_neutron_l3/models/netconf_yang/class_map.py:
```diff
@@ -0,0 +1,106 @@
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from collections import OrderedDict
```
Do we really need `OrderedDict`? I think this was only used because `xmltodict` did so. With python3.8+ we have insertion order anyway.

sebageek commented on asr1k_neutron_l3/models/netconf_yang/class_map.py:
```diff
@@ -0,0 +1,106 @@
+from asr1k_neutron_l3.models.netconf_yang.ny_base import NyBase
+
+
+class ClassMapConstants():
```
We don't need the `()` here.

sebageek commented on asr1k_neutron_l3/models/netconf_yang/class_map.py:
```diff
@@ -0,0 +1,106 @@
+
+class ClassMap(NyBase):
+    ID_FILTER = """
+                <native>
```
All other `ID_FILTER`s have the `<native xmlns="http://cisco.com/ns/yang/Cisco-IOS-XE-native"` `xmlns`. I think we should keep that consistent.

sebageek commented on asr1k_neutron_l3/models/netconf_yang/class_map.py:
```diff
@@ -0,0 +1,106 @@
+                </native>
+                """
+
+    LIST_KEY = None
```
Do we need this to be set? I mean I would expect this to be the default, but APPARENTLY the default is `""` set in NyBase. Which means that the driver will try to unpack an empty argument. So maybe we should've set this to `None` in NyBase, still? Did you run into a problem with this?

swagner-de replied:
Yes, I did run into exactly what you described. There's no list key for this type and with the defaults I get empty responses. Most of our other classes deriving from `NyBase` define a `LIST_KEY`. Route Maps also set it to `None`.

sebageek replied:
We can investigate this in a follow up (to the asr1k driver, not your patch)

sebageek commented on asr1k_neutron_l3/models/netconf_yang/class_map.py:
```diff
@@ -0,0 +1,106 @@
+                return uuid
+
+    def is_orphan(self, all_fwaas_external_policies, *args, **kwargs):
+        if self.policy_id:
+            return self.policy_id not in all_fwaas_external_policies
+        return False
```
[very optional] `return self.policy_id and self.policy_id not in all_fwaas_external_policies`
This doesn't need a check for the all_router_ids stuff because it's not listed in the default device cleaner stuff, right?

swagner-de replied:
Exactly, so far this is only called from the fwaas-cleaner.

sebageek commented on asr1k_neutron_l3/models/netconf_yang/l3_interface.py:
```diff
@@ -183,11 +195,34 @@ def to_dict(self, context):
+            }
+
+        redundancy = OrderedDict()
+        if self.rii and self.redundancy_group and self.redundancy_group_decrement:
```
I presume these only make sense together, even when coming from the device, right?

swagner-de replied:
Yes, they are configured in a single statement and none of them is optional.

sebageek commented on asr1k_neutron_l3/models/netconf_yang/ny_base.py:
```diff
@@ -826,7 +826,7 @@ def _internal_validate(self, context, should_be_none=False):
     @execute_on_pair(result_type=DiffResult)
     def _validate(self, context, should_be_none=False):
-        return self._internal_validate(context=context, should_be_none=False)
+        return self._internal_validate(context=context, should_be_none=should_be_none)
```
Interesting! Have you observed any effect for the existing codebase on this?

swagner-de replied:
[This is the only place](https://github.com/sapcc/asr1k-neutron-l3/blob/stable/yoga-m3/asr1k_neutron_l3/models/neutron/l3/bgp.py#L59) were it can be `True`. So my expectation is that we might have executed updates on BGP too often.

sebageek commented on asr1k_neutron_l3/models/netconf_yang/parameter_map.py:
```diff
@@ -0,0 +1,118 @@
+from collections import OrderedDict
+from asr1k_neutron_l3.models.netconf_yang import xml_utils
+from asr1k_neutron_l3.models.netconf_yang.ny_base import NyBase
+from asr1k_neutron_l3.common import utils
```
Newline between importblocks

sebageek commented on asr1k_neutron_l3/models/netconf_yang/parameter_map.py:
```diff
@@ -0,0 +1,118 @@
+        _C = ParameterMapConstants
+        content = OrderedDict()
+
+        content[_C.ID] = self.id
```
[optional/info] Heh, got me for a second. have been searching for `self.vrf` but NyBase apparently automatically maps this to `self.id`. Interesting!

sebageek commented on asr1k_neutron_l3/models/netconf_yang/service_policy.py:
```diff
@@ -0,0 +1,146 @@
+
+    @classmethod
+    def remove_wrapper(cls, content, context):
+        content = super()._remove_base_wrapper(content, context)
```
If you don't override `_remove_base_wrapper()` you can just call `cls._remove_base_wrapper(content, context)`

sebageek commented on asr1k_neutron_l3/models/netconf_yang/service_policy.py:
```diff
@@ -0,0 +1,146 @@
+    def _wrapper_preamble(self, content, context):
+        policy = OrderedDict()
+        policy[ServicePolicyConstants.POLICY] = content
+        return policy
```
[optional] shorter
```python
    return {ServicePolicyConstants.POLICY: content}
```

sebageek commented on asr1k_neutron_l3/models/netconf_yang/service_policy.py:
```diff
@@ -0,0 +1,146 @@
+        content = OrderedDict()
+
+        content[xml_utils.NS] = xml_utils.NS_CISCO_POLICY
+        # On update, YANG might have a different state of the order of class-maps than the CLI, hence we do a PUT
```
Would it make sense to sort the class maps as well so we are always in a stable order or does the ordering already happen somewhere else (haven't reviewed the DB part yet)?

swagner-de replied:
[For us the order is static](https://github.com/sapcc/asr1k-neutron-l3/blob/fwaas-support/asr1k_neutron_l3/models/neutron/l3/firewall.py#L122-L126), it's always: _match this acl_, _drop_. Also I wouldn't have an attribute to order on deep down in this yang class like a sequence or so.

sebageek replied:
In our call you told me it's already ordered from neutron side and with device side the order should be retained by device (at least I'd expect that).

sebageek commented on asr1k_neutron_l3/models/netconf_yang/service_policy.py:
```diff
@@ -0,0 +1,146 @@
+        content[_C.TYPE] = self.type
+        if len(self.classes) > 0:
+            content[_C.CLASS] = list()
+        for class_ in self.classes:
+            content[_C.CLASS].append(class_.to_dict(context))
```
If this is only executed when there are classes then indent into the already existing `if`.

swagner-de replied:
Thought the goal is to always save indention levels ;)

sebageek commented on asr1k_neutron_l3/models/netconf_yang/service_policy.py:
```diff
@@ -0,0 +1,146 @@
+        content = OrderedDict()
+
+        content[xml_utils.NS] = xml_utils.NS_CISCO_POLICY
+        # On update, YANG might have a different state of the order of class-maps than the CLI, hence we do a PUT
```
Where's the PUT though? ;) I think this is a copypaste comment error.

swagner-de replied:
Jep.

sebageek commented on asr1k_neutron_l3/models/netconf_yang/zone.py:
```diff
@@ -0,0 +1,74 @@
+
+    @property
+    def neutron_router_id(self):
+        if self.id is not None and self.id.startswith(FWAAS_ZONE_PREFIX):
```
I think I added this before, but I'd regard a `if self.id and self.id.startswith(...):` as shorter / more readable, as  `""` would also result in `False` for the startswith anyway.

sebageek commented on asr1k_neutron_l3/models/neutron/l3/access_list.py:
```diff
@@ -20,7 +20,10 @@
 class AccessList(base.Base):
     def __init__(self, id, drop_on_17_3=False):
-        super(AccessList, self).__init__()
+        # If we use super(AccessList, self) here the MRO of type(self) is determining the __init__
+        # we call. If self is a child, that could lead to not calling base.Base.__init__().
+        # If the second argument to super is a type, the returned method is unbound, hence we pass self to it.
+        super(AccessList, AccessList).__init__(self)
```
Where does this happen? Where is this causing a problem? What type is the `self`?

swagner-de replied:
Jep, the problem comes from  [`firewall.AccessList`](https://github.com/sapcc/asr1k-neutron-l3/blob/fwaas-support/asr1k_neutron_l3/models/neutron/l3/firewall.py#L53) which inherits from `FirewallPolicyObject` but also from this class.  `FirewallPolicyObject` in turn also  inherits from `base.Base` again.  So a diamond shaped inheritance.
Say `firewall.AccessList.__init__()` calls to `access_list.AccessList.__init__()`, now I want to prevent that the `super()` in line 26 calls `FirewallPolicyObject.__init__()` but instead `base.Base().__init__()`

I am very open to other suggestions here.

sebageek commented on asr1k_neutron_l3/models/neutron/l3/access_list.py:
```diff
@@ -30,9 +33,43 @@ def _rest_definition(self):
-                                           protocol=rule.protocol, ipv4_address=rule.source, mask=rule.source_mask,
-                                           dest_ipv4_address=rule.destination, dest_mask=rule.destination_mask)
+
+            ip_args = dict()
```
`{}` maybe? :)

sebageek commented on asr1k_neutron_l3/models/neutron/l3/access_list.py:
```diff
@@ -50,11 +87,17 @@ def append_rule(self, rule):
 
 
 class Rule(object):
```
[optional] while you're at it, remove the `(object)`, it's the default with python3.

sebageek commented on asr1k_neutron_l3/models/neutron/l3/access_list.py:
```diff
@@ -30,9 +33,43 @@ def _rest_definition(self):
-                                           protocol=rule.protocol, ipv4_address=rule.source, mask=rule.source_mask,
-                                           dest_ipv4_address=rule.destination, dest_mask=rule.destination_mask)
+
+            ip_args = dict()
+            if rule.source_mask:
+                ip_args['ipv4_address'] = rule.source
+                ip_args['mask'] = rule.source_mask
+            else:
+                ip_args['host'] = rule.source
+
+            if rule.destination_mask:
+                ip_args['dest_ipv4_address'] = rule.destination
+                ip_args['dest_mask'] = rule.destination_mask
+            else:
+                ip_args['dst_host'] = rule.destination
```
With some of the other `models/neutron/` classes we have helper methods that are called in `__init__()` and set an attribute that is then used as arguments to the `_rest_definition`. Maybe we could do this here as well? It would reduce the complexity of this long method and also give us the benefit of having the input args as part of the model, which often helps debugging it. Same could be done for `port_args`. Maybe even just a `build_rules()` method setting rules. What do you think?

swagner-de replied:
I have know hidden the `port_args` and `ip_args` creation in the `Rule` object. This could all be a lot simpler if `netconf_yang.access_list.ACERule` and `ACLRule` would not need the `id` aka the ACL name. I believe they also do not need it, but I see that more to be the task of a follow up.

sebageek commented on asr1k_neutron_l3/models/neutron/l3/access_list.py:
```diff
@@ -30,9 +33,43 @@ def _rest_definition(self):
+                        port_args[yang_direction + '_eq'] = ports[0]
+                    else:
+                        port_args[yang_direction + '_range1'] = ports[0]
+                        port_args[yang_direction + '_range2'] = ports[1]
```
`f`-string maybe?

sebageek commented on asr1k_neutron_l3/models/neutron/l3/access_list.py:
```diff
@@ -30,9 +33,43 @@ def _rest_definition(self):
+            if rule.protocol == 'tcp' and rule.established:
+                port_args['established'] = True
+
+            if rule.protocol not in {'tcp', 'udp'} and rule.named_message_type:
```
[optional] Tuple instead of set for tcp and udp?

sebageek commented on asr1k_neutron_l3/models/neutron/l3/access_list.py:
```diff
@@ -30,9 +33,43 @@ def _rest_definition(self):
+                port_args['established'] = True
+
+            if rule.protocol not in {'tcp', 'udp'} and rule.named_message_type:
+                port_args['named_message_type'] = rule.named_message_type
```
Do we have any protection against unknown protocols or something like that? Is that done by the API?

swagner-de replied:
I only ever saw the `named_message_type` property filled with icmp, specifying the ICMP type message. However the FWaaS API does not allow anything more specific than `icmp`. Consequently this field is currently only written to by `firewall.AccessList.MIMIC_STATEFUL_RULES`.

sebageek replied:
It's only set by us, got it! :)

sebageek commented on asr1k_neutron_l3/models/neutron/l3/access_list.py:
```diff
@@ -50,11 +87,17 @@ def append_rule(self, rule):
 class Rule(object):
-    def __init__(self, action='permit', protocol='ip', source=None, source_mask=None,
-                 destination=None, destination_mask=None):
+    def __init__(self, action='permit', protocol='ip',
```
Do these default parameters make sense? Do we need to provide `action` by default?

swagner-de replied:
In the code I have written, I explicitly pass `action`  and `protocol` to the constructor. However the pbr and NAT code would need a change if we drop the default arguments. I mainly kept it this way to not influence other code.

sebageek commented on asr1k_neutron_l3/models/neutron/l3/firewall.py:
```diff
@@ -0,0 +1,264 @@
+
+class FirewallPolicyObject(base.Base):
+
+    PREFIX = 'OBJ-FWAAS-'
```
Why not set this to `None` and complain if the inheriting class has not set it?

sebageek commented on asr1k_neutron_l3/models/neutron/l3/firewall.py:
```diff
@@ -0,0 +1,264 @@
+    def get_id_by_policy_id(cls, policy_id: str) -> str:
+        if cls.PREFIX == FirewallPolicyObject.PREFIX:
+            raise NotImplementedError("Class derived from 'FirewallPolicyObject' must define static var 'PREFIX'")
+        return "{}{}".format(cls.PREFIX, policy_id)
```
`f`-string maybe?

sebageek commented on asr1k_neutron_l3/models/neutron/l3/firewall.py:
```diff
@@ -0,0 +1,264 @@
+    def __init__(self, policy_id: str):
+        self.policy_id = policy_id
+        self.id = self.get_id_by_policy_id(self.policy_id)
+        super().__init__()
```
I think this is at the top for all others, any reason that is not the case here?

swagner-de replied:
Because the methods before are `@classmethods`

sebageek commented on asr1k_neutron_l3/models/neutron/l3/firewall.py:
```diff
@@ -0,0 +1,264 @@
+    ACTIONS = {
+            'allow': 'permit',
+            'deny': 'deny',
+            'reject': 'deny'
```
`,` after last entry, just if we get more actions.

sebageek commented on fwaas_ios_config.MD:
````diff
@@ -0,0 +1,223 @@
+ ip route vrf 6051ee469c184bdca168a746ae0c69f6 0.0.0.0 0.0.0.0 10.237.208.1
+```
+
+As customers expect statefull behavior we prepend every customer ACL with entries that mimic statefulness. Hence we allow all established connections and ICMP packets that would elsewise require a state to be kept. As we map OpenStack firewall policy objects to an IOS ACL, we would like to be able to use the same ACL in a class-map for _ZBF_ as well.
````
s/elsewise/otherwise/

sebageek commented on asr1k_neutron_l3/models/neutron/l3/router.py:
```diff
@@ -422,6 +487,9 @@ def _delete(self):
 
         results.append(self.vrf.delete())
 
+        # We do not delete fwaas acls here as the acl could
+        # still be in use by an another router
```
we could do an `is_orphan()` check and then do it, or do you think the periodic cleaner is suited better?

swagner-de replied:
Yes, but not without asking the neutron server for more information or putting that information straight into `sync_data`. The ladder one I feel like we would be more susceptible for race conditions. 
I think there's no harm in leaving them here and have them deleted later by a cleanup loop.

sebageek commented on asr1k_neutron_l3/models/neutron/l3/router.py:
```diff
@@ -85,6 +88,8 @@ def __init__(self, router_info):
                            rd=self.router_atts.get('rd'), routable_interface=self.routable_interface,
                            rt_import=self.rt_import, rt_export=self.rt_export, global_vrf_id=global_vrf_id)
 
+        self._build_fwaas_conf(router_info)
```
Most of the other methods use `self.router_info` instead of the passed argument. Shall we follow that convention?

sebageek commented on asr1k_neutron_l3/models/neutron/l3/router.py:
```diff
@@ -292,6 +316,35 @@ def _build_prefix_lists(self):
         return result
 
+    def _build_fwaas_conf(self, router_info):
+        self.fwaas_conf = list()
```
`list()` vs `[]`. Also nearly all other helper methods generate an item, a list or dict themselves and then return that. Shall we do it the same way here?

sebageek commented on asr1k_neutron_l3/models/neutron/l3/router.py:
```diff
@@ -136,10 +141,29 @@ def _build_interfaces(self):
 
         inf_ports = self.router_info.get('_interfaces', [])
+
+        # turn that dict around, we now want to know which policy should be bound on which interface in
+        # which direction
+        port_policy_map = defaultdict(dict)
+        for policy_id, policy in self.router_info.get('fwaas_policies', {}).items():
+            for egress_port in policy['egress_ports']:
+                port_policy_map[egress_port]['egress'] = policy_id
+            for ingress_port in policy['ingress_ports']:
+                port_policy_map[ingress_port]['ingress'] = policy_id
+
         if inf_ports is not None:
             for inf_port in inf_ports:
+                firewall_acls = dict()
+                ingress = port_policy_map[inf_port['id']].get('ingress')
+                egress = port_policy_map[inf_port['id']].get('egress')
+                if ingress:
+                    firewall_acls['ingress_acl'] = firewall.AccessList.get_id_by_policy_id(ingress)
+                if egress:
+                    firewall_acls['egress_acl'] = firewall.AccessList.get_id_by_policy_id(egress)
+
                 interfaces.append(
-                    l3_interface.InternalInterface(self.router_id, inf_port, self._port_extra_atts(inf_port)))
+                    l3_interface.InternalInterface(self.router_id, inf_port, self._port_extra_atts(inf_port),
+                                                   **firewall_acls))
```
This increases the complexity of the method quite a bit. Could we maybe make it shorter and/or more concise? We could maybe move the dict-creation into an extra method. If the other options you come up with don't increase readability we can also keep it this way.

This could also be a use case for the `:=` operator :)
```python
firewall_acls = {}
if ingress := port_policy_map[inf_port['id']].get('ingress'):
    firewall_acls['ingress_acl'] = firewall.AccessList.get_id_by_policy_id(ingress)
...
```

swagner-de replied:
I have packed this into an own method and rewritten it slightly with a few less lines.

sebageek commented on asr1k_neutron_l3/models/neutron/l3/router.py:
```diff
@@ -292,6 +316,35 @@ def _build_prefix_lists(self):
+            # We also want to link the VRF to a policer so we limit VRFs (by boilerplate)
+            self.fwaas_conf.append(firewall.FirewallVrfPolicer(self.router_id))
+            # Mark all interfaces for stateful firewalling
+            for interface in self.interfaces.all_interfaces:
+                interface.has_stateful_firewall = True
```
Isn't this only needed for the gateway interface? I see that you have code checking this Flag in `InternalInterface`, but there this attribute is only used in the constructor, which has already been executed once you set it from this method.

swagner-de replied:
And here I was thinking _He's right, but why does it still work then?_. Turns out it only works for `GatewayInterface`s and I need to read up if we also need it on `InternalInterfaces`, but I think we will.

swagner-de replied:
Ok, this needs to be on all Interfaces. I'll fix up the `InternalInterface` class.

sebageek commented on asr1k_neutron_l3/models/neutron/l3/router.py:
```diff
@@ -353,6 +406,18 @@ def _update(self):
+
+        # If there are no external policies, we can create dangling objects which will trigger the deletion
+        if not self.fwaas_external_policies['ingress'] and not self.fwaas_external_policies['egress']:
+            results.append(firewall.DanglingZonePairExtIngress(self.router_id).update())
+            results.append(firewall.DanglingZonePairExtEgress(self.router_id).update())
+            results.append(firewall.DanglingFirewallVrfPolicer(self.router_id).update())
+            results.append(firewall.DanglingZone(self.router_id).update())
```
Why do we need the `Dangling` zones here? Couldn't we just call `delete()` of the original Zone?

swagner-de replied:
For instance the zone pairs need their service policy to be set to `None` when being deleted. However its constructor would set to the default allow policy. Yes, I could handle all of that here in the code or alter the constructor of the regular class to handle that. Never the less I liked the way to hide it all in an extra class, even if the `Dangling*` class often times really doesn't do anything different I enjoyed the consitency to just create dangling objects and be done with it.

sebageek commented on asr1k_neutron_l3/models/neutron/l3/router.py:
```diff
@@ -488,6 +556,20 @@ def diff(self):
+                diff_results[obj.id] = d.to_dict()
+
+        if not self.fwaas_external_policies['ingress'] and not self.fwaas_external_policies['egress']:
+            for obj in [firewall.DanglingZonePairExtIngress(self.router_id),
+                        firewall.DanglingZonePairExtEgress(self.router_id),
+                        firewall.DanglingFirewallVrfPolicer(self.router_id),
+                        firewall.DanglingZone(self.router_id)]:
```
Why do we need the Dangling ones here, couldn't we use the original ones?

swagner-de replied:
See my other answer regarding that topic

sebageek commented on asr1k_neutron_l3/models/neutron/l3/interface.py:
```diff
@@ -173,19 +186,31 @@ def _nat_address(self):
+                              access_group_out=self.egress_acl,
+                              access_group_in=self.ingress_acl)
+
+        if self.has_stateful_firewall:
+            interface_args['redundancy_group'] = 1
+            interface_args['redundancy_group_decrement'] = 1
+            interface_args['rii'] = self.bridge_domain
```
As written in `router.py` this flag will never be true, as it is only set _after_ `__init__()` has been called.

swagner-de replied:
True. Fixed.

sebageek commented on asr1k_neutron_l3/models/neutron/l3/interface.py:
```diff
@@ -173,19 +186,31 @@ def _nat_address(self):
-                                            secondary_ip_addresses=self.secondary_ip_addresses,
-                                            nat_inside=True, redundancy_group=None, route_map="pbr-{}".format(self.vrf),
-                                            ntp_disable=True, arp_timeout=cfg.CONF.asr1k_l3.internal_iface_arp_timeout)
+                      f'network:{self.router_port["network_id"]};subnet:{self._primary_subnet_id}')
```
This looks like an accidental dedent by one space.

sebageek commented on asr1k_neutron_l3/plugins/l3/rpc/rpc_api.py:
```diff
@@ -93,6 +93,14 @@ def get_deleted_router_atts(self, context, **kwargs):
 
         return router_atts
 
+    @instrument()
+    def get_policies_on_agent(self, context, host=None, only_external=False):
+        return self.db.get_policies_on_agent(context, host, only_external=only_external)
+
+    @instrument()
+    def get_routers_with_policy(self, context, host=None, policy_id=None, only_external=False):
+        return self.db.get_routers_with_policy(context, host, policy_id, only_external=only_external)
+
```
Is `host=None` a valid default argument or is this always used?

swagner-de replied:
It's not of course. I was blindsided by the other methods in this class.

sebageek commented on asr1k_neutron_l3/plugins/db/asr1k_db.py:
```diff
@@ -77,10 +79,72 @@ class DBPlugin(db_base_plugin_v2.NeutronDbPluginV2,
         super(DBPlugin, self).__init__()
+    
+
+    def get_router_id_by_ports(self, context, ports):
```
I think this should be named `get_router_ids_by_ports()` as it can return multiple router ids.

sebageek commented on asr1k_neutron_l3/plugins/db/asr1k_db.py:
```diff
@@ -77,10 +79,72 @@ class DBPlugin(db_base_plugin_v2.NeutronDbPluginV2,
+    
+
+    def get_router_id_by_ports(self, context, ports):
+        query = context.session.query(models_v2.Port.device_id) \
```
I think `neutron.db.models.l3.RouterPort` might already have the relationship you're looking for aka `SELECT router_id FROM routerports WHERE port_id IN ...`.

sebageek commented on asr1k_neutron_l3/plugins/db/asr1k_db.py:
```diff
@@ -77,10 +79,72 @@ class DBPlugin(db_base_plugin_v2.NeutronDbPluginV2,
+            .join(fwaas.FirewallGroup, or_(
+                        fwaas.FirewallGroup.egress_firewall_policy_id == fwaas.FirewallPolicy.id,
+                        fwaas.FirewallGroup.ingress_firewall_policy_id == fwaas.FirewallPolicy.id)) \
+            .join(fwaas.FirewallGroupPortAssociation) \
+            .join(models_v2.Port) \
```
How to these two behave when you don't add a condition for joining? Do they have a default filter?

swagner-de replied:
Usually sqlalchemy examines the foreign key relationship, but since we have multiple joins here, I'll make it explicit.

sebageek commented on asr1k_neutron_l3/plugins/db/asr1k_db.py:
```diff
@@ -77,10 +79,72 @@ class DBPlugin(db_base_plugin_v2.NeutronDbPluginV2,
+            .filter(fwaas.FirewallGroup.admin_state_up is True) \
+            .filter(models_v2.Port.device_owner.in_([
+                *n_constants.ROUTER_INTERFACE_OWNERS, n_constants.DEVICE_OWNER_ROUTER_GW])) \
+            .filter(models_v2.Port.device_id == router_id) \
```
Also an option to use routerports table again, if you like.

sebageek commented on asr1k_neutron_l3/plugins/db/asr1k_db.py:
```diff
@@ -77,10 +79,72 @@ class DBPlugin(db_base_plugin_v2.NeutronDbPluginV2,
+                        fwaas.FirewallGroup.ingress_firewall_policy_id == fwaas.FirewallPolicy.id)) \
+            .join(fwaas.FirewallGroupPortAssociation) \
+            .join(models_v2.Port) \
+            .filter(fwaas.FirewallGroup.admin_state_up is True) \
```
I'd use `== True`, `is` generally checks identity... But if both work, I'd at least keep it consistent with below.

sebageek commented on asr1k_neutron_l3/plugins/db/asr1k_db.py:
```diff
@@ -77,10 +79,72 @@ class DBPlugin(db_base_plugin_v2.NeutronDbPluginV2,
+            .join(fwaas.FirewallGroup, or_(fwaas.FirewallGroup.egress_firewall_policy_id == fwaas.FirewallPolicy.id,
+                                           fwaas.FirewallGroup.ingress_firewall_policy_id == fwaas.FirewallPolicy.id)) \
+            .join(fwaas.FirewallGroupPortAssociation, fwaas.FirewallGroupPortAssociation.firewall_group_id == fwaas.FirewallGroup.id) \
+            .join(models_v2.Port, models_v2.Port.id == fwaas.FirewallGroupPortAssociation.port_id) \
+            .join(l3agent_models.RouterL3AgentBinding,
+                  l3agent_models.RouterL3AgentBinding.router_id == models_v2.Port.device_id) \
```
Might be a candidate for routerports, though I see you need the `device_owner` below, so... maybe not or this would need an extra join? Not sure if it's generally nicer to not depend on`device_id` if we can use the explicit routerports table.

swagner-de replied:
Luckily [that's also in `RouterPorts`](https://github.com/sapcc/neutron/blob/stable/yoga-m3/neutron/db/models/l3.py#L35-L40). Good substitute.

sebageek commented on asr1k_neutron_l3/neutron/services/service_drivers/asr1k/fwaas_driver.py:
```diff
@@ -0,0 +1,153 @@
+
+class ASR1KFWaaSDriver(driver_api.FirewallDriverDB, driver_api.FirewallDriverRPCMixin):
+
+    SUPPORTED_L3_PORTS = set([nl_constants.DEVICE_OWNER_ROUTER_GW, nl_constants.DEVICE_OWNER_ROUTER_INTF])
```
[optional] Could be a direct set without list by doing `foo = {1, 2, 3}`

sebageek commented on asr1k_neutron_l3/neutron/services/service_drivers/asr1k/fwaas_driver.py:
```diff
@@ -0,0 +1,153 @@
+        LOG.debug("No rpc listener available atm, so not starting one")
+        return []
+
+    def notify_asr1k_agent(self, entity, id, context, router_ids=None, fwg_policy_ids=None):
```
[optional] might be a private method (add a `_`)

sebageek commented on asr1k_neutron_l3/neutron/services/service_drivers/asr1k/fwaas_driver.py:
```diff
@@ -0,0 +1,153 @@
+            for policy_id in fwg_policy_ids:
+                # build a unique list of policies and agents with an router that contains the ACL
+                # notification_directory = {(policy, agent) = router_id }
+                notification_directory = {(policy_id, x.host): x.device_id
+                                          for x in self.asr1k_db.get_routers_with_policy(context, policy_id=policy_id)}
```
if we have multiple routers with the same host the last one wins. That's the explicit goal here, right?

swagner-de replied:
It doesn't really matter, does it? I need to iterate over the whole list anyways in order to not forget an agent.

sebageek replied:
So you just need any random router (which happens to always be the last from the list fetched from neutron-server), got it.

sebageek commented on asr1k_neutron_l3/neutron/services/service_drivers/asr1k/fwaas_driver.py:
```diff
@@ -0,0 +1,153 @@
+        if isinstance(router_ids, list):
+            self.notifier.routers_updated(context, router_ids)
+            LOG.debug("{} {} has port/policy associated that belongs to routers {}."
+                      " Notifying asr1k agent.".format(entity, id, router_ids))
```
Would replace that with normal lot formatting `%s`

sebageek commented on asr1k_neutron_l3/neutron/services/service_drivers/asr1k/fwaas_driver.py:
```diff
@@ -0,0 +1,153 @@
+        if not firewall_group['admin_state_up']:
+            return
+        if firewall_group['ingress_firewall_policy_id'] is None and firewall_group['egress_firewall_policy_id'] is None:
+            LOG.debug("firewall_group {} has no policies attached. Doing nothing".format(firewall_group['id']))
```
Log formatting

sebageek commented on asr1k_neutron_l3/neutron/services/service_drivers/asr1k/fwaas_driver.py:
```diff
@@ -0,0 +1,153 @@
+                LOG.debug("%s %s has changed, policy %s associated. Notifying agent to call update on"
+                          " router_id %s on agent %s", entity, id, policy_id, router_id, agent)
+                if isinstance(router_ids, list):
+                    router_ids.append(router_id)
```
+ `if router_id not in router_ids` to check if the router id we're trying to udpate is already scheduled externally

sebageek commented on asr1k_neutron_l3/neutron/services/service_drivers/asr1k/fwaas_driver.py:
```diff
@@ -0,0 +1,153 @@
+        policies_changed = \
+            old_firewall_group['ingress_firewall_policy_id'] != new_firewall_group['ingress_firewall_policy_id'] \
+            or \
+            old_firewall_group['egress_firewall_policy_id'] != new_firewall_group['egress_firewall_policy_id']
```
[optional] If you want less`\` you could write `()` around the whole statement and feel freeeeeee!

sebageek commented on asr1k_neutron_l3/neutron/services/service_drivers/asr1k/fwaas_driver.py:
```diff
@@ -0,0 +1,153 @@
+    def delete_firewall_group_postcommit(self, context, firewall_group):
+        if not firewall_group['admin_state_up']:
+            return
+        if firewall_group['ingress_firewall_policy_id'] is None and firewall_group['egress_firewall_policy_id'] is None:
```
is the `is None` check necessary? Could this be `if not (firewall_group['ingress...'] or firewall_group['egress...'])`?

sebageek commented on asr1k_neutron_l3/neutron/services/service_drivers/asr1k/fwaas_driver.py:
```diff
@@ -0,0 +1,153 @@
+        router_ids = self.asr1k_db.get_router_id_by_ports(context, firewall_group['ports'])
+        self.notify_asr1k_agent('firewall_group', firewall_group['id'], context, router_ids=router_ids)
+
+    # Firewall Policy
+    # create hooks omitted as action needs only to be taken once a policy is bound to a port via a group
+    # delete hooks omitted as policies cannot be deleted when they are bound to a firewall group
```
[optional] could be a docstring

swagner-de replied:
Where would you put it? In `update_firewall_policy_postcommit`?

sebageek commented on asr1k_neutron_l3/neutron/services/service_drivers/asr1k/fwaas_driver.py:
```diff
@@ -0,0 +1,153 @@
+        action_attributes = ['protocol', 'ip_version', 'source_ip_address', 'destination_ip_address',
+                             'source_port', 'destination_port', 'action', 'enabled']
+        for attr in action_attributes:
+            if old_firewall_rule[attr] != new_firewall_rule[attr]:
```
`if any(old_firewall_rule[attr] != new_firewall_rule[attr] for attrin action_attributes):`

return probably not needed.

sebageek commented on asr1k_neutron_l3/plugins/l3/agents/device_cleaner.py:
```diff
@@ -71,6 +77,39 @@ def _get_all_router_ids(self):
+        """
+
+        LOG.debug("Fetching FWaaS objects from device")
+        fwaas_classes = [ServicePolicy, ClassMap, AccessList]
```
[optional] Could keep this as a class variable like we have with `ENTITIES` as for example `FWAAS_ENTITIES`. Looks a bit more consistent, also if we ever need it from the outside it's accessible.

sebageek commented on asr1k_neutron_l3/plugins/l3/agents/device_cleaner.py:
```diff
@@ -71,6 +77,39 @@ def _get_all_router_ids(self):
+
+        LOG.debug("Fetching FWaaS objects from device")
+        fwaas_classes = [ServicePolicy, ClassMap, AccessList]
+        device_objects = {}
```
As we already know what entries we'll be having here we could explicitly write `device_objects = {device for device in ASR1KPair().contexts]`

sebageek commented on asr1k_neutron_l3/plugins/db/asr1k_db.py:
```diff
@@ -77,10 +79,72 @@ class DBPlugin(db_base_plugin_v2.NeutronDbPluginV2,
                ):
     def __init__(self):
         super(DBPlugin, self).__init__()
+    
```
Please remove the extra whitespace + newline

sebageek commented on asr1k_neutron_l3/plugins/db/asr1k_db.py:
```diff
@@ -77,10 +79,72 @@ class DBPlugin(db_base_plugin_v2.NeutronDbPluginV2,
+            .filter(models_v2.Port.id.in_(ports)) \
+            .filter(models_v2.Port.device_owner.in_([
+                *n_constants.ROUTER_INTERFACE_OWNERS, n_constants.DEVICE_OWNER_ROUTER_GW])) \
+            .distinct()    
```
trailing whitespace

sebageek commented on asr1k_neutron_l3/models/neutron/l3/firewall.py:
```diff
@@ -0,0 +1,264 @@
+            # check if there is an IP address/CIDR for each direction
+            # if so do the whole mask, wildcard dance
+            for direction in ('source', 'destination'):
+                if rule[direction + '_ip_address'] is not None:
```
[optional] Would use `f`strings here

sebageek commented on asr1k_neutron_l3/models/neutron/l3/firewall.py:
```diff
@@ -0,0 +1,264 @@
+    @property
+    def _rest_definition(self):
+        return ncClassMap(id=self.id, acl_id=AccessList.get_id_by_policy_id(self.policy_id),
+                        type='inspect', prematch='match-all')
```
indent +2 spaces

sebageek commented on asr1k_neutron_l3/models/neutron/l3/firewall.py:
```diff
@@ -0,0 +1,264 @@
+                        type='inspect', prematch='match-all')
+
+
+class DanglingClassMap(ClassMap):
+
+    def update(self):
+        return self.delete()
```
Why do we need an extra class here, why can't the user of this just create a `ClassMap` and call `delete()` on it?

swagner-de replied:
It's not needed, but some of the classes here need some additional processing and I preferred it to be consistent for all objects. I also find it sleeker to hide it in the object.
I know we are missing this consistency in the driver code so far, but I can still keep it consistent for all the FWaaS objects...

swagner-de replied:
I just noticed that you might asked why we specifically need `DanglingClassMap` and as a matter of fact. That's nowhere used. The other `Dangling*` objects are however.

sebageek commented on asr1k_neutron_l3/models/neutron/l3/firewall.py:
```diff
@@ -0,0 +1,264 @@
+
+class FirewallZoneObject(base.Base):
+
+    PREFIX = 'OBJ-FWAAS-'
```
Don't we have this as a constant somewhere?

sebageek commented on asr1k_neutron_l3/models/neutron/l3/firewall.py:
```diff
@@ -0,0 +1,264 @@
+
+    @classmethod
+    def get_id_by_router_id(cls, router_id: str) -> str:
+        if cls.PREFIX == FirewallZoneObject.PREFIX:
```
If this is a baseclass I'd use `None` as the value and check for unset.

sebageek commented on asr1k_neutron_l3/models/neutron/l3/firewall.py:
```diff
@@ -0,0 +1,264 @@
+    def get_id_by_vrf(cls, vrf: str) -> str:
+        rid = utils.vrf_id_to_uuid(vrf)
+        if not isinstance(rid, str):
+            raise ValueError(f"VRF {vrf} could not be converted to router id.")
```
You could also write `if not rid`. Technically the method returns a bool if it couldn't be converted but anything evaluating to false wouldn't be acceptible here anyway.

sebageek commented on asr1k_neutron_l3/models/neutron/l3/firewall.py:
```diff
@@ -0,0 +1,264 @@
+        if self.policy_id is not None:
+            self.service_policy = ServicePolicy.get_id_by_policy_id(self.policy_id)
+        else:
+            self.service_policy = self.DEFAULT_ALLOW_INSPECT_POLICY
```
What is this default policy? I might be missing some background knowledge here.

swagner-de replied:
It's the policy that we set if no policy was given. It generally a _allow all_ but with stateful inspection. So we will allow everything if we see a proper handshake.

sebageek commented on asr1k_neutron_l3/models/neutron/l3/firewall.py:
```diff
@@ -0,0 +1,264 @@
+    def vrf(self) -> str:
+        return utils.uuid_to_vrf_id(self.router_id)
+
+    @property
+    def id(self) -> str:
+        return self.vrf
```
Why do we need this alias?

swagner-de replied:
I think I took this over because `NyBase` does a lot of back and forth mapping between `neutron_router_id` and `id` but I believe in this higher level class it is not needed, at least I couldn't grep any usage of this.

swagner-de replied:
Found it [here](https://github.com/sapcc/asr1k-neutron-l3/pull/106/commits/36cd00c8af0a57614d5ff0ec031ab56cfaaab02a#diff-a0bf0323d4849fe859ec3ea5c7a3ddad500ed5667ca4638af927dbc4cea71bf4R568). Using it to put the diff results into a dict, as a unique identifier.

sebageek replied:
Link not working anymore, but I'm going to trust you on this :)

sebageek reviewed (commented):
I like it, cool implementation. Looking forward of having this in prod.

swagner-de committed:
Reorganize Access List rule creation

To remove clutter from the AccessList class we move all the code that
generates the attributes for the corresponding yang class into `Rule`.

swagner-de committed:
Resolve diamond shaped inheritance

In the previous implementation `firewall.AccessList` would inherit from
`l3.AccessList` which inherited from `base.Base`.
`firewall.AccessList` also inherited from
`firewall.FirewallPolicyObject`. That lead to a Diamond Shaped
inheritance. The fact that `FirewallPolicyObject` and `l3.AccessList`
had constructors with different arguments made it difficult to let
`super()` call the constructors with different arguments.

To resolve that, the diamond Shaped inheritance was resolved making it
clearer which `__init__`'s `super()` calls which superclasses
constructor.

The different signatures of the constructor still pose a problem, hence
we send all supplemental arguments of a constructor to `**kwargs`.

swagner-de committed:
Remove Dangling* classes as they don't really do anything

swagner-de committed:
Make `is_orphan` in subclasses generic

All netconf_yang classes inherit their `is_orphan` from BulkOperations.
Each of these subclasses need different lists to determine if the
classes object is orphan or not. Let's simplify adding arguments to this
method by using keyword arguments. All subclasses shall send unnecessary
arguments to `*args, **kwargs` while only `BulkOperations.is_orphan`
should know about the whole argument list.

swagner-de committed:
Avoid using FWaaS database queries when firewall_v2 is not enabled

sven-rosenzweig reviewed (approved):
Nice work.
